### PR TITLE
Use T-API for critical parts of blobFromImagesWithParams

### PR DESCRIPTION
**Pull request by kallaballa**
Pertaining Issue: https://github.com/opencv/opencv/issues/5697

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by vpisarev** [comment]
@kallaballa, thank you for the pull request. I have some doubts regarding it.

1. it references the issue https://github.com/opencv/opencv/issues/5697 that complains about zero performance gain when using OpenCV + T-API for inference. However, a typical deep learning model includes a few dozens if not 100+ layers, so the preprocessing, which is just a few operations.
2. in the experimental inference engine https://github.com/vpisarev/ficus/tree/master/lib/NN, which we slowly, part by part, migrate to OpenCV, there is a single-pass "blobFromImage" (it's called differently there) preprocessing operation (https://github.com/vpisarev/ficus/blob/master/lib/NN/Preprocess.fx) - I would prefer to migrate it and optimize using OpenCL rather than to use multi-pass collection of primitive operations.
3. If the whole pipeline is to capture high-resolution images/video stream, downscale it and run a deep net, then resize might be one of the bottleneck. In such a case user could manually copy the input to UMat, call OpenCV's resize, download result to CPU (.getMat()) and then run the rest of preprocessing on a downscaled image.
4. It would be nice to do a little research and provide numbers - by how much this patch improves performance on different hardware: 1. desktop CPU with discrete GPU, 2. mobile Intel/AMD CPU with integrated graphics, 3. ARM with Mali or another OpenCL-capable graphics.
5. If the performance, as measured in item 4, is not noticeably better than with CPU (it could well be the case, see item 1), I would prefer to keep preprocessing on CPU. From time to time we get various problem with our OpenCL kernels. I would like to keep things more stable and predictable, unless there is clear advantage.
6. In OpenCV 5 we plan, among other things, to make OpenCV much more friendly to various GPU/NPU accelerators. UMat should become universal data structure for buffers, directly accessible by various acceleration APIs, including OpenCL, Vulkan, GLSL, CUDA etc. Then we can get back to such a patch. But, as I said in 2., I would prefer the whole blobFromImage(s) to be converted into a single well-optimized parallel loop. Then it could be ported to OpenCL/GLSL/CUDA etc. and then user can directly call blobFromImages and pass UMat there. Now it's not the proper time yet, in my opinion.

**Comment by kallaballa** [comment]
Alright. At least I'll provide some numbers based on my machine.

**Comment by kallaballa** [comment]
I modified [one of the V4D demos](https://viel-zu.org/opencv/doxygen/html/d3/ddf/v4d_beauty.html) to track face detection time using ```getTickCount()```.
**My machine:** 11th Gen Intel(R) Core(TM) _i7-1160G7_ @ 1.20GHz with _Iris Xe Graphics_
**Video used (from 00:01:00):** https://www.youtube.com/watch?v=hUAT8Jm_dvw

In this scenario 1000 iterations of FaceDetectorYN::detect take:
* With patch: 25.7823s
* Without patch: 34.0648s

I'll add flame graphs.

I understand your arguments will but given the simplicity of the patch the risk/gain ratio isn't so bad :)

**Comment by kallaballa** [comment]
Flame graphs zoomed in on ```FaceDetectorYN::detect```
Without patch:
fg_without

With patch:
fg_with

**Comment by kallaballa** [comment]
Btw. I made similar patches e.g for TrackerKCF with considerable performance gain. I guess I can drop those?

**Comment by vpisarev** [comment]
@kallaballa, thank you for the quick response. The acceleration in your case is noticeable, indeed! What's the resolution of images that you feed to blobFromImagesWithParams?

I think, I can propose a compromise solution that will make everybody happy.
Let's implement "compute follow data" principle (that is already used in OpenCV when it comes to T-API):
if inputs are UMat's (images.kind == _InputArray::STD_VECTOR_UMAT), use your new branch, otherwise use the original one.

1. OpenCV video capture can return the frames in UMat's, not just Mat's, and some video capture backends support UMat output in very efficient way.
2. You can use .getUMat() function to convert each processed image into UMat quickly.
3. Yes, users will have to modify their apps to use your branch, but for now such explicit behaviour for this for the better.

**Comment by kallaballa** [comment]
> @kallaballa, thank you for the quick response. The acceleration in your case is noticeable, indeed! What's the resolution of images that you feed to blobFromImagesWithParams?

960x540
 
> I think, I can propose a compromise solution that will make everybody happy. Let's implement "compute follow data" principle (that is already used in OpenCV when it comes to T-API): if inputs are UMat's (images.kind == _InputArray::STD_VECTOR_UMAT), use your new branch, otherwise use the original one.
> 
> 1. OpenCV video capture can return the frames in UMat's, not just Mat's, and some video capture backends support UMat output in very efficient way.
> 2. You can use .getUMat() function to convert each processed image into UMat quickly.
> 3. Yes, users will have to modify their apps to use your branch, but for now such explicit behaviour for this for the better.

Sounds good. Will implement it like that.

**Comment by kallaballa** [comment]
I think that's it. 00804cb

**Comment by kallaballa** [comment]
> I think that's it. [00804cb](https://github.com/opencv/opencv/commit/00804cb0e6e3ef59c2fe02d35cca76b71b34caa3)

If you are alright with the general approach, i'd improve the implementation a bit more.

**Comment by kallaballa** [comment]
Also I wrote a test that compares detection. there are differences (some frames not detected with UMat) I am trying to track down.

File: modules/dnn/src/dnn_utils.cpp [hunk 1]
```diff
@@ -45,17 +45,18 @@ void getMat(Mat& blob, InputArray blob_, AccessFlag flag) {
...
-void makeMatFromBlob(Mat& m, InputArray blob, int i, int j, int rows, int cols, int type) {
+void getChannelFromBlob(Mat& m, InputArray blob, int i, int j, int rows, int cols, int type) {
     m = Mat(rows, cols, type, blob.getMat().ptr(i, j));
 }
 
-void makeMatFromBlob(UMat& m, InputArray blob, int i, int j, int rows, int cols, int type) {
+void getChannelFromBlob(UMat& m, InputArray blob, int i, int j, int rows, int cols, int type) {
     UMat ublob = blob.getUMat();
-    int offset = i * cols + j;
+    int offset = (i * ublob.step.p[0] + j * ublob.step.p[1]) / ublob.elemSize();
```
Comment (kallaballa) [inline modules/dnn/src/dnn_utils.cpp]: I forgot to take into account step() and elemSize() for the offset. Now it works

**Comment by kallaballa** [comment]
I have put other more code on the fast-path by porting NaryEltwiseLayer to UMat. I understand given the developments around 5.0 that this doesn't have priority but there is considerable gain. Should i post figures and make a PR? Only part left to port is ResizeLayer to keep it on the GPU all the time.

**Comment by kallaballa** [comment]
Here a flame graph of the current state:
flame

**Review by vpisarev — APPROVED** [review]

File: modules/dnn/src/dnn_utils.cpp [hunk 2]
```diff
@@ -34,8 +76,13 @@ void blobFromImage(InputArray image, OutputArray blob, double scalefactor,
...
     CV_TRACE_FUNCTION();
-    std::vector<Mat> images(1, image.getMat());
-    blobFromImages(images, blob, scalefactor, size, mean, swapRB, crop, ddepth);
+    if (image.kind() == _InputArray::UMAT) {
+        std::vector<UMat> images(1, image.getUMat());
+        blobFromImages(images, blob, scalefactor, size, mean, swapRB, crop, ddepth);
+    } else if (image.kind() == _InputArray::MAT) {
+        std::vector<Mat> images(1, image.getMat());
+        blobFromImages(images, blob, scalefactor, size, mean, swapRB, crop, ddepth);
+    }
```
Comment (opencv-alalek) [inline modules/dnn/src/dnn_utils.cpp:85]: `else` do NOOP - this is wrong.

Only `UMAT` and `STD_VECTOR_UMAT` should be handled in dedicated code path.
All other handling should not be changed by this patch.

Comment (kallaballa) [reply modules/dnn/src/dnn_utils.cpp:85]: I understand

File: modules/dnn/src/dnn_utils.cpp [hunk 3]
```diff
@@ -232,6 +278,73 @@ void blobFromImagesWithParams(InputArrayOfArrays images_, OutputArray blob_, con
...
     else
         CV_Error(Error::StsUnsupportedFormat, "Unsupported data layout in blobFromImagesWithParams function.");
+
+    CV_Assert(blob_.total());
+}
+
+void blobFromImagesWithParams(InputArrayOfArrays images, OutputArray blob, const Image2BlobParams& param) {
+    CV_TRACE_FUNCTION();
+
+        if (images.kind() == _InputArray::STD_VECTOR_UMAT) {
```
[the remark this answers is not in the document]
Comment (kallaballa) [reply modules/dnn/src/dnn_utils.cpp:288]: ic

File: modules/dnn/src/dnn_utils.cpp [hunk 4]
```diff
@@ -21,6 +21,48 @@ Image2BlobParams::Image2BlobParams(const Scalar& scalefactor_, const Size& size_
...
+
+void getVector(InputArrayOfArrays images_, std::vector<UMat>& images) {
+    images_.getUMatVector(images);
+}
+
+void getMat(UMat& blob, InputArray blob_, AccessFlag flag) {
+    if(blob_.kind() == _InputArray::UMAT)
+        blob = blob_.getUMat();
+    else if(blob_.kind() == _InputArray::MAT) {
+        blob = blob_.getMat().getUMat(flag);
```
[the remark this answers is not in the document]
Comment (kallaballa) [reply modules/dnn/src/dnn_utils.cpp:36]: alright

[the remark this answers is not in the document]
Comment (kallaballa) [reply modules/dnn/src/dnn_utils.cpp:36]: so clone?

File: modules/dnn/src/dnn_utils.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kallaballa) [reply modules/dnn/src/dnn_utils.cpp]: [e7107e7](https://github.com/opencv/opencv/pull/23894/commits/e7107e734c26f754afc8dfc8da55e98184288bc3)

File: modules/dnn/src/dnn_utils.cpp (hunk 2 above)
Comment (kallaballa) [reply modules/dnn/src/dnn_utils.cpp:85]: [86095e6](https://github.com/opencv/opencv/pull/23894/commits/86095e6fb3ea563e643b9f3496dbaaedaeac0d57)

File: modules/dnn/src/dnn_utils.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (opencv-alalek) [reply modules/dnn/src/dnn_utils.cpp]: `.clone()` is overkill.

`blob_.getMat()` return temporary object - it should be alive till `.getUMat(flag)` release.

---

We don't need this method at all as there is ` _InputArray::getUMat()`: https://github.com/opencv/opencv/blob/4.8.0/modules/core/src/matrix_wrap.cpp#L126C5-L126C27

[the remark this answers is not in the document]
Comment (kallaballa) [reply modules/dnn/src/dnn_utils.cpp]: Thx!

**Comment by asmorkalov** [comment]
```
modules/dnn/src/dnn_utils.cpp:298: tab in indent.
+	if (images.kind() == _InputArray::STD_VECTOR_UMAT) {
modules/dnn/src/dnn_utils.cpp:299: tab in indent.
+		if(blob.kind() == _InputArray::UMAT) {
modules/dnn/src/dnn_utils.cpp:300: tab in indent.
+			UMat& u = blob.getUMatRef();
modules/dnn/src/dnn_utils.cpp:301: tab in indent.
+			blobFromImagesWithParamsImpl<cv::UMat>(images, u, param);
modules/dnn/src/dnn_utils.cpp:302: tab in indent.
+			return;
modules/dnn/src/dnn_utils.cpp:303: tab in indent.
+		} else if(blob.kind() == _InputArray::MAT) {
modules/dnn/src/dnn_utils.cpp:304: tab in indent.
+			UMat u = blob.getMatRef().getUMat(ACCESS_WRITE);
modules/dnn/src/dnn_utils.cpp:305: tab in indent.
+			blobFromImagesWithParamsImpl<cv::UMat>(images, u, param);
modules/dnn/src/dnn_utils.cpp:306: tab in indent.
+			u.copyTo(blob);
modules/dnn/src/dnn_utils.cpp:307: tab in indent.
+			return;
modules/dnn/src/dnn_utils.cpp:308: tab in indent.
+		}
modules/dnn/src/dnn_utils.cpp:309: tab in indent.
+	} else if (images.kind() == _InputArray::STD_VECTOR_MAT) {
modules/dnn/src/dnn_utils.cpp:310: tab in indent.
+		if(blob.kind() == _InputArray::UMAT) {
modules/dnn/src/dnn_utils.cpp:311: tab in indent.
+			Mat m = blob.getUMatRef().getMat(ACCESS_WRITE);
modules/dnn/src/dnn_utils.cpp:312: tab in indent.
+			blobFromImagesWithParamsImpl<cv::Mat>(images, m, param);
modules/dnn/src/dnn_utils.cpp:313: tab in indent.
+			m.copyTo(blob);
modules/dnn/src/dnn_utils.cpp:314: tab in indent.
+			return;
modules/dnn/src/dnn_utils.cpp:315: tab in indent.
+		} else if(blob.kind() == _InputArray::MAT) {
modules/dnn/src/dnn_utils.cpp:316: tab in indent.
+			Mat& m = blob.getMatRef();
modules/dnn/src/dnn_utils.cpp:317: tab in indent.
+			blobFromImagesWithParamsImpl<cv::Mat>(images, m, param);
modules/dnn/src/dnn_utils.cpp:318: tab in indent.
+			return;
modules/dnn/src/dnn_utils.cpp:319: tab in indent.
+		}
modules/dnn/src/dnn_utils.cpp:320: tab in indent.
+	}
```

**Comment by kallaballa** [comment]
fbe2ccb

File: modules/objdetect/src/face_detect.cpp [hunk 5]
```diff
@@ -102,12 +102,21 @@ class FaceDetectorYNImpl : public FaceDetectorYN
...
-        Mat input_blob = dnn::blobFromImage(pad_image);
 
+        Mat input_blob;
+        if(input_image.kind() == _InputArray::UMAT) {
+            // Pad input_image with divisor 32
+            UMat pad_image;
+            padWithDivisor(input_image, pad_image);
+            // Build blob from input image
+            input_blob = dnn::blobFromImage(pad_image);
+        } else {
```
Comment (opencv-alalek) [inline modules/objdetect/src/face_detect.cpp:113]: @vpisarev T-API declares what we should not have such code separation on "user" side.

**Comment by asmorkalov** [comment]
@opencv-alalek @vpisarev Is it ready for merge?

**Comment by kallaballa** [comment]
Yay!

**Pull request merged**